src/SubHask/Algebra/Matrix.hs

```diff
@@ -296,3 +308,206 @@ instance Category (Matrix' vect r) where
+    ones = undefined
+
+instance
+  (P.Fractional (A.Exp r), VectorSpace r, Prim r, ValidACCMatrix b v m n r) =>
+  VectorSpace (ACCMatrix b v (n::Symbol) (m::Symbol) r) where
+    {-# INLINE (./) #-}
+    (./)  (ACCMatrix v) r = ACCMatrix(A.map (\x -> x A./ (A.constant r)) v)
+    {-# INLINE (./.) #-}
+    (./.) (ACCMatrix v1) (ACCMatrix v2) = ACCMatrix(A.zipWith (A./) v1 v2)
+
+instance
+  (ValidACCMatrix b v m n r, Monoid r, ValidLogic r, Prim r, IsScalar r)
+  => IxContainer (ACCMatrix b v (n::Symbol) (m::Symbol) r) where
+
+  -- {-# INLINE (!) #-}
+  -- (!) (ACCMatrix m) i = let
+  --   l = A.size m
+  --   in ACCMatrix m!!(i `div` l, i `mod` l)
+
+-- {-# INLINE row #-}
+-- row :: (ValidACCMatrix b v m n r
+--   ) => ACCMatrix b v (n::Symbol) (m::Symbol) r -> Int -> vect
+--   row m i = A.slice m (A.Z A.:. (i::Int) A.:. A.All)
+--
+-- {-# INLINE col #-}
+-- col ::
+--   ( ValidACCMatrix b v m n r
+--   ) => ACCMatrix b v (n::Symbol) (m::Symbol) r -> Int -> vect
+-- col m j = A.slice m (A.Z A.:. A.All A.:. (j::Int))
+--
+--
+-- --taken from http://www.cse.unsw.edu.au/~chak/papers/repa.pdf
+-- {-# INLINE mmult #-}
+-- mmult ::
+--   ( ValidACCMatrix b v m n r
+--   )
+--   => ACCMatrix b v (n::Symbol) (x0::Symbol) r
+--   -> ACCMatrix b v (x0::Symbol) (m::Symbol) r
+--   -> ACCMatrix b v (n::Symbol) (m::Symbol) r
+-- mmult arr brr = A.sum (A.zipWith (*) arrRepl brrRepl)
+--   where
+--     trr = A.transpose brr
+--     arrRepl = A.replicate (A.Z A.:. A.All A.:. colsB A.:. A.All) arr
+--     brrRepl = A.replicate (A.Z A.:. rowsA A.:. A.All A.:. A.All) trr
+--     (A.Z A.:. colsA A.:. rowsA) = A.shape arr
+--     (A.Z A.:. colsB A.:. rowsB) = A.shape brr
+--
+-- {-# INLINE transpose #-}
+-- transpose ::
+--   ( ValidACCMatrix b v m n r
+--   )
+--   => ACCMatrix b v (m::Symbol) (n::Symbol) r
+--   -> ACCMatrix b v (m::Symbol) (n::Symbol) r
+-- transpose m = A.transpose m
+
+data ACCMatrix' b v (m::Symbol) (n::Symbol) r where
+    ACCZero ::
+        (ValidACCMatrix b v m n r) =>
+        ACCMatrix' b v (m::Symbol) (n::Symbol) r
+
+    ACCId ::
+        (ValidACCMatrix b v m n r) =>
+        {-#UNPACK#-}!(Scalar r) -> ACCMatrix' b v (m::Symbol) (n::Symbol) r
+
+    ACCMat ::
+        (ValidACCMatrix b v m n r) =>
+        {-#UNPACK#-}!(ACCMatrix b v (m::Symbol) (n::Symbol) r)
+        -> ACCMatrix' b v (m::Symbol) (n::Symbol) r
+
+type instance Scalar (ACCMatrix' b v (m::Symbol) (n::Symbol) r) = Scalar r
+type instance Logic (ACCMatrix' b v (m::Symbol) (n::Symbol) r) = Bool
+
+type instance ACCMatrix' b v (m::Symbol) (n::Symbol) r  >< n =
+  ACCTensor_Linear (ACCMatrix' b v (m::Symbol) (n::Symbol) r ) n
+type family ACCTensor_Linear m n where
+    ACCTensor_Linear (ACCMatrix' b v (m::Symbol) (n::Symbol) r) c =
+      ACCMatrix' b v (m::Symbol) (n::Symbol) r
+
+-- Categor failes with:
+--       The first argument of ‘Category’ should have kind ‘k0 -> k0 -> *’,
+--         but ‘ACCMatrix' b v (m :: Symbol) (n :: Symbol) r’ has kind ‘*’
```

Comment:
You need to reorder the paremeters.  The last two parameters should be the `m` and `n`, and `r` should be before them.

You can look at the `Matrix` type for an example.  It has a similar set of parameters in a different order.